server/server.go

```diff
@@ -0,0 +1,257 @@
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+	"go.mongodb.org/mongo-driver/x/mongo/driver/connstring"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/health"
+	"google.golang.org/grpc/health/grpc_health_v1"
+)
+
+const (
+	configPort                  = "port"
+	envPrefix 					= "FVS"
+	configMongoConnectionString = "mongo_host"
+	configMongoClientConnectionTimeout = "mongo_client_connection_timeout"
+	configMongoClientPingTimeout       = "mongo_client_ping_timeout"
+	configElasticAPMIgnoreURLS         = "elastic_apm_ignore_urls"
+
+
+)
+
+func init() {
+	viper.SetDefault(configPort, "8080")
+	viper.SetDefault(configElasticAPMIgnoreURLS, "/grpc.health.v1.Health/Check")
+	viper.SetDefault(configMongoConnectionString, "mongodb://localhost:27017/favorite")
+	viper.SetDefault(configMongoClientConnectionTimeout, 10)
+	viper.SetDefault(configMongoClientPingTimeout, 10)
+	viper.SetEnvPrefix(envPrefix)
+	viper.AutomaticEnv()
+}
+
+// FavoriteServer is a structure that holds the permission grpc server and its services and configuration. 
+type FavoriteServer struct {
+	pb.UnimplementedFavoriteServer
+	*grpc.Server
+	logger 		*logrus.Logger
+	port 		string
+	healthCheckInterval int
+	favoriteService service.Service
+}
+
+
+// Serve accepts incoming connections on the listener `lis`, creating a new
+// ServerTransport and service goroutine for each. The service goroutines
+// read gRPC requests and then call the registered handlers to reply to them.
+// Serve returns when `lis.Accept` fails with fatal errors. `lis` will be closed when
+// this method returns.
+// If `lis` is nil then Serve creates a `net.Listener` with "tcp" network listening
+// on the configured `TCP_PORT`, which defaults to "8080".
+// Serve will return a non-nil error unless Stop or GracefulStop is called.
+func (s FavoriteServer) Serve(lis net.Listener) {
+	listener := lis
+	if lis ==nil {
+		l, err := net.Listen("tcp", ":"+s.port)
+		if err != nil {
+			s.logger.Fatalf("failed to listen: %v", err)
+		}
+
+		listener = l
+	}
+
+	log.Printf("listening and serving grpc server on port %s", s.port)
+	if err := s.Server.Serve(listener); err != nil {
+		s.logger.Fatalf(err.Error())
+	}
+
+}
+
+
+// NewServer configures and creates a grpc.Server instance
+// health check service.
+// Configure using environment variables.
+// `HEALTH_CHECK_INTERVAL`: Interval to update serving state of the health check server.
+// `PORT`: TCP port on which the grpc server would serve on.
+func NewServer(logger *logrus.Logger) *FavoriteServer {
+	if logger == nil {
+		logger = ilogger.NewLogger()
+	}
+
+	// Set up grpc server opts with logger interceptor.
+	serverOpts := append(
+		serverLoggerInterceptor(logger),
+		grpc.MaxRecvMsgSize(16<<20),
+	)
+
+	grpcServer := grpc.NewServer(
+		serverOpts...,
+	)
+
+	controller, err := initMongoDBController(viper.GetString(configMongoConnectionString))
```

Comment:
You do not need to send parameters to the function. You can also access viper variables from the initMongoDBController function.

Reply:
done